llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -5209,6 +5209,26 @@ PerformBUILD_VECTORCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
   return DAG.getNode(ISD::BITCAST, DL, VT, PRMT);
 }
 
+static SDValue combineADDRSPACECAST(SDNode *N,
```

Comment:
This will potentially hide invalid casts between specific AS'es.
E.g. `ASC(shared->local->shared)` or `ASC(shared->local->generic`) will all of a sudden start working.

While we may be allowed to do so, because it's the input that's invalid, hiding such errors may not be the best option.
I'd prefer to narrow the folding to apply only to the valid casts, and leave invalid ones as is -- they will be handled/or not when we try lowering individual casts and, at least, will still be present in the DAG in their original form.

Reply:
Fair enough, I've removed the transitive case for now.

Reply:
I'm not against transitive casts in principle. If we have a valid combinations, collapsing them is fine. It's just PTX does not give us that much wiggle room, as we can only cast to generic and back, and we can't use that for casts between different AS, so A->B->A will already handle them.

The only possibility seems to be casts between shared::cta and shared::cluster. PTX docs are not quite clear on that (or I didn't find the right place in the docs yet). If you can clarify whether casts between shared variants are legal, then we could restore transitive cast folding for them.

Reply:
I agree there aren't non-bug cases where this fold would be useful. The only case I can imagine would be identifying invalid casts in code that will never execute allowing us to simplify by propagating undef. 

> The only possibility seems to be casts between shared::cta and shared::cluster. PTX docs are not quite clear on that (or I didn't find the right place in the docs yet). If you can clarify whether casts between shared variants are legal, then we could restore folding of transitive casts via generic AS for them.

Even if this cast were legal, we'd need to go through the generic AS to represent this in PTX, so folding it here would likely just make our lives harder. Further we don't represent shared::cta and shared::cluster as distinct AS making this a moot point.